lib/Sema/TypeCheckStmt.cpp

```diff
@@ -1246,8 +1246,20 @@ Stmt *StmtChecker::visitBraceStmt(BraceStmt *BS) {
 
       bool hadTypeError = TC.typeCheckExpression(SubExpr, DC, TypeLoc(),
                                                  CTP_Unused, options);
-      if (isDiscarded && !hadTypeError)
+
+      // If a closure expression is unused, the user might have intended
+      // to write "do { ... }".
+      if (auto *CE = dyn_cast<ClosureExpr>(SubExpr)) {
+        TC.diagnose(CE->getLoc(), diag::expression_unused_closure);
```

Comment:
Thanks for pointing this out. I also found that my use of `cast<>` was crashing because I failed to account for CaptureListExpr.